// graphile-config/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolvePresets = exports.resolvePreset = exports.isResolvedPreset = exports.Middleware = exports.AsyncHooks = exports.applyHooks = exports.orderedApply = void 0;
exports.sortedPlugins = sortedPlugins;
require("./interfaces.js");
const sort_ts_1 = require("./sort.js");
var functionality_ts_1 = require("./functionality.js");
Object.defineProperty(exports, "orderedApply", { enumerable: true, get: function () { return functionality_ts_1.orderedApply; } });
var hooks_ts_1 = require("./hooks.js");
Object.defineProperty(exports, "applyHooks", { enumerable: true, get: function () { return hooks_ts_1.applyHooks; } });
Object.defineProperty(exports, "AsyncHooks", { enumerable: true, get: function () { return hooks_ts_1.AsyncHooks; } });
var middleware_ts_1 = require("./middleware.js");
Object.defineProperty(exports, "Middleware", { enumerable: true, get: function () { return middleware_ts_1.Middleware; } });
var resolvePresets_ts_1 = require("./resolvePresets.js");
Object.defineProperty(exports, "isResolvedPreset", { enumerable: true, get: function () { return resolvePresets_ts_1.isResolvedPreset; } });
Object.defineProperty(exports, "resolvePreset", { enumerable: true, get: function () { return resolvePresets_ts_1.resolvePreset; } });
Object.defineProperty(exports, "resolvePresets", { enumerable: true, get: function () { return resolvePresets_ts_1.resolvePresets; } });
function sortedPlugins(plugins) {
    if (plugins) {
        return (0, sort_ts_1.sortWithBeforeAfterProvides)(plugins, "name");
    }
    else {
        return [];
    }
}
