// @graphiql/react/dist/index.js
/* empty css               */
import { ArgumentIcon, ChevronDownIcon, ChevronLeftIcon, ChevronUpIcon, CloseIcon, CopyIcon, DeprecatedArgumentIcon, DeprecatedEnumValueIcon, DeprecatedFieldIcon, DirectiveIcon, DocsFilledIcon, DocsIcon, EnumValueIcon, FieldIcon, HistoryIcon, ImplementsIcon, KeyboardShortcutIcon, MagnifyingGlassIcon, MergeIcon, PenIcon, PlayIcon, PlusIcon, PrettifyIcon, ReloadIcon, RootTypeIcon, SettingsIcon, StarFilledIcon, StarIcon, StopIcon, TrashIcon, TypeIcon } from "./icons/index.js";
import { KEY_MAP, formatShortcutForOS, isMacOs } from "./constants.js";
import { useCopyQuery, useEditorContext, useEditorStore, useExecutionContext, useExecutionStore, useMergeQuery, usePluginContext, usePluginStore, usePrettifyEditors, useSchemaContext, useSchemaStore, useStorage, useStorageContext, useTheme } from "./deprecated.js";
import { useMonaco } from "./stores/monaco.js";
import { cleanupDisposables } from "./utility/cleanup-disposables.js";
import { createBoundedUseStore } from "./utility/create-bounded-use-store.js";
import { createEditor, getOrCreateModel, onEditorContainerKeyDown } from "./utility/create-editor.js";
import { debounce } from "./utility/debounce.js";
import { formatJSONC, parseJSONC, tryParseJSONC } from "./utility/jsonc.js";
import { markdown } from "./utility/markdown.js";
import { URI } from "monaco-editor/esm/vs/base/common/uri.js";
import { KeyMod } from "monaco-editor/esm/vs/editor/common/services/editorBaseApi.js";
import { KeyCode } from "monaco-editor/esm/vs/editor/common/standalone/standaloneEnums.js";
import { Range } from "monaco-editor/esm/vs/editor/common/core/range.js";
import { pick } from "./utility/pick.js";
import { useDragResize } from "./utility/resize.js";
import { clsx } from "clsx";
import { useChangeHandler, useDidUpdate, useEditorState, useHeadersEditorState, useOperationsEditorState, useOptimisticState, useVariablesEditorState } from "./utility/hooks.js";
import { ExecuteButton } from "./components/execute-button/index.js";
import { ToolbarButton } from "./components/toolbar-button/index.js";
import { ToolbarMenu } from "./components/toolbar-menu/index.js";
import { RequestHeadersEditor } from "./components/request-headers-editor.js";
import { ImagePreview } from "./components/image-preview.js";
import { GraphiQLProvider, useGraphiQL, useGraphiQLActions } from "./components/provider.js";
import { OperationEditor } from "./components/operation-editor.js";
import { ResponseEditor } from "./components/response-editor.js";
import { VariablesEditor } from "./components/variables-editor.js";
import { Button, UnStyledButton } from "./components/button/index.js";
import { ButtonGroup } from "./components/button-group/index.js";
import { Dialog } from "./components/dialog/index.js";
import { DropdownMenu } from "./components/dropdown-menu/index.js";
import { MarkdownContent } from "./components/markdown-content/index.js";
import { Spinner } from "./components/spinner/index.js";
import { Tab, Tabs } from "./components/tabs/index.js";
import { Tooltip } from "./components/tooltip/index.js";
import { Root } from "@radix-ui/react-visually-hidden";
export {
  ArgumentIcon,
  Button,
  ButtonGroup,
  ChevronDownIcon,
  ChevronLeftIcon,
  ChevronUpIcon,
  CloseIcon,
  CopyIcon,
  DeprecatedArgumentIcon,
  DeprecatedEnumValueIcon,
  DeprecatedFieldIcon,
  Dialog,
  DirectiveIcon,
  DocsFilledIcon,
  DocsIcon,
  DropdownMenu,
  EnumValueIcon,
  ExecuteButton,
  FieldIcon,
  GraphiQLProvider,
  RequestHeadersEditor as HeaderEditor,
  HistoryIcon,
  ImagePreview,
  ImplementsIcon,
  KEY_MAP,
  KeyCode,
  KeyMod,
  KeyboardShortcutIcon,
  MagnifyingGlassIcon,
  MarkdownContent,
  MergeIcon,
  PenIcon,
  PlayIcon,
  PlusIcon,
  PrettifyIcon,
  OperationEditor as QueryEditor,
  Range,
  ReloadIcon,
  ResponseEditor,
  RootTypeIcon,
  SettingsIcon,
  Spinner,
  StarFilledIcon,
  StarIcon,
  StopIcon,
  Tab,
  Tabs,
  ToolbarButton,
  ToolbarMenu,
  Tooltip,
  TrashIcon,
  TypeIcon,
  UnStyledButton,
  URI as Uri,
  VariablesEditor as VariableEditor,
  Root as VisuallyHidden,
  cleanupDisposables,
  clsx as cn,
  createBoundedUseStore,
  createEditor,
  debounce,
  formatJSONC,
  formatShortcutForOS,
  getOrCreateModel,
  isMacOs,
  markdown,
  onEditorContainerKeyDown,
  parseJSONC,
  pick,
  tryParseJSONC,
  useChangeHandler,
  useCopyQuery,
  useDidUpdate,
  useDragResize,
  useEditorContext,
  useEditorState,
  useEditorStore,
  useExecutionContext,
  useExecutionStore,
  useGraphiQL,
  useGraphiQLActions,
  useHeadersEditorState,
  useMergeQuery,
  useMonaco,
  useOperationsEditorState,
  useOptimisticState,
  usePluginContext,
  usePluginStore,
  usePrettifyEditors,
  useSchemaContext,
  useSchemaStore,
  useStorage,
  useStorageContext,
  useTheme,
  useVariablesEditorState
};
